genie-web/src/main/java/com/netflix/genie/web/agent/apis/rpc/v4/endpoints/GRpcAgentFileStreamServiceImpl.java

```diff
@@ -142,333 +151,630 @@ public GRpcAgentFileStreamServiceImpl(
-                pendingTransferBuffer.closeForError(t);
+        private synchronized void updateManifestAndStream(
+            final ControlStreamObserver controlStreamObserver,
+            final String jobId,
+            final DirectoryManifest manifest
+        ) {
+            // Keep the most recent manifest for each job id
+            this.manifestCache.put(jobId, manifest);
+
+            // Keep the most recent control stream for each job id
+            final ControlStreamObserver previousObserver = this.controlStreamMap.put(jobId, controlStreamObserver);
+            if (previousObserver != null && previousObserver != controlStreamObserver) {
+                // If the older one is still present, close it
+                previousObserver.closeStreamWithError(
+                    new IllegalStateException("A new stream was registered for the same job id: " + jobId)
+                );
             }
 
-            final StreamBuffer inProgressTransferBuffer = this.inProgressTransferBuffersMap.remove(streamId);
-            if (inProgressTransferBuffer != null) {
-                inProgressTransferBuffer.closeForError(t);
-            }
+            this.registry.gauge(MANIFEST_CACHE_SIZE_GAUGE, this.manifestCache.estimatedSize());
+            this.registry.gauge(CONTROL_STREAMS_GAUGE, this.controlStreamMap.size());
         }
-    }
 
-    private void handleFileTransferCompletion(
-        final FileTransferStreamObserver fileTransferStreamObserver,
-        @Nullable final String streamId
-    ) {
-        log.info("Completed file transfer: {}", streamId);
-        this.pendingTransferObserversSet.remove(fileTransferStreamObserver);
+        private synchronized void removeControlStream(
+            final ControlStreamObserver controlStreamObserver,
+            @Nullable final Throwable t
+        ) {
+            log.debug("Control stream {}", t == null ? "completed" : "error: " + t.getMessage());
 
-        final StreamBuffer pendingTransferBuffer = this.pendingTransferBuffersMap.remove(streamId);
-        if (pendingTransferBuffer != null) {
-            pendingTransferBuffer.closeForCompleted();
+            final boolean foundAndRemoved = this.controlStreamMap
+                .entrySet()
+                .removeIf(entry -> entry.getValue() == controlStreamObserver);
+
+            if (foundAndRemoved) {
+                log.debug(
+                    "Removed a control stream due to {}",
+                    t == null ? "completion" : "error: " + t.getMessage()
+                );
+            }
         }
 
-        final StreamBuffer inProgressTransferBuffer = this.inProgressTransferBuffersMap.remove(streamId);
-        if (inProgressTransferBuffer != null) {
-            inProgressTransferBuffer.closeForCompleted();
+        @Override
+        public synchronized void contribute(final Info.Builder builder) {
+            builder
+                .withDetail("file_control_streams", this.controlStreamMap.size())
```

Comment:
Keys should probably be constants. Also since this renders as JSON in browser I think it's preferred to used `-` as delimiter instead of `_`

Reply:
Oh, i intentionally avoided `-` because i vaguely remember it's a problem for JS.